### PR TITLE
Add command line interface for hashlink library.

### PR DESCRIPTION
**Pull request by msporny**

### CODE REVIEW

File: bin/hl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  .option('-u, --url <location>', 'target URL for the hashlink')
+  .option('-l, --legacy', 'create a hashlink for a legacy URL')
+  .option('-h, --hash <algorithm>', 'cryptographic hash algorithm', 'sha2-256')
+  .option('-b, --base <encoding>', 'base encoding', 'base58-btc')
+  .action(async (dataFile, options) => {
+    try {
+      let data, urls, transforms, meta = undefined;
+      // read data from disk if provided
+      if(dataFile) {
+        data = fs.readFileSync(dataFile, null);
```
Comment (msporny) [inline bin/hl]: This is adding an extra null byte to the end of the buffer, which is wrong. Thoughts @dlongley @davidlehn?

File: Hashlink.js [hunk 2]
```diff
@@ -64,8 +64,14 @@ export class Hashlink {
     meta = {...meta, url: urls};
 
     // generate the encoded cryptographic hash
-    const outputData = await transforms.reduce(async (output, transform) =>
-      this.registeredTransforms[transform].encode(await output), data);
+    const outputData = await transforms.reduce(async (output, transform) => {
+      const transformer = this.registeredTransforms[transform];
+      if(transformer === undefined) {
+        throw new Error(`Unknown transformation ${transform}`)
```
Comment (davidlehn) [inline Hashlink.js]: Maybe quote the name:
```suggestion
        throw new Error(`Unknown transformation "${transform}"`)
```

**Comment by davidlehn** [comment]
Add a "[bin](https://docs.npmjs.com/files/package.json#bin)" rule in package.json:
```json
  "bin": {
    "hl": "./bin/hl.js"
  }
```

File: Hashlink.js (hunk 2 above)
Comment (davidlehn) [reply Hashlink.js]: Or maybe the style here is colon before unquoted user data?

Comment (dlongley) [reply Hashlink.js]: The above suggestion is good except for a missing period at the end.

File: bin/hl (hunk 1 above)
Comment (msporny) [reply bin/hl]: Nevermind, it was a stray `\n`, fixed in tests.

**Comment by msporny** [comment]
> Add a "[bin](https://docs.npmjs.com/files/package.json#bin)" rule in package.json:
> 
> ```json
>   "bin": {
>     "hl": "./bin/hl.js"
>   }
> ```

Fixed in 2aa78980130058e93b0485f57005eda622114f18

File: Hashlink.js [hunk 3]
```diff
@@ -64,8 +64,14 @@ export class Hashlink {
...
 
     // generate the encoded cryptographic hash
-    const outputData = await transforms.reduce(async (output, transform) =>
-      this.registeredTransforms[transform].encode(await output), data);
+    const outputData = await transforms.reduce(async (output, transform) => {
+      const transformer = this.registeredTransforms[transform];
+      if(transformer === undefined) {
+        throw new Error(`Unknown transformation "${transform}".`);
+      }
+
```
Comment (dlongley) [inline Hashlink.js:72]:
```suggestion
```

File: bin/hl [hunk 4]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env node
+
+const fs = require('fs-extra');
+const program = require('commander');
+const hl = require('../');
+const {URL} = require('url');
+
+function collect(value, previous) {
+  return previous.concat([value]);
+}
```
Comment (dlongley) [inline bin/hl:10]: Move helpers to bottom of the file.

File: bin/hl [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+          const outputUrl = new URL(inputUrl);
+          outputUrl.searchParams.set('hl', rawHashlink);
+          console.log(outputUrl.toString());
+        });
+      } else {
+        // print out the hashlink
+        console.log(hashlink);
+      }
+    } catch(e) {
+      console.error(`${e}`);
```
Comment (dlongley) [inline bin/hl:47]:
```suggestion
      console.error(e);
```

File: bin/hl [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        }
+        // check to see if hashlink is valid
+        const valid = await hl.verify({data, hashlink});
+        if(valid) {
+          console.log('hashlink is valid');
+        } else {
+          throw new Error('hashlink is invalid');
+        }
+      } catch(e) {
+        console.error(`${e}`);
```
Comment (dlongley) [inline bin/hl:85]:
```suggestion
        console.error(e);
```

**Review by dlongley — APPROVED** [review]

**Pull request merged**